File: imagepicker/src/main/java/com/nguyenhoanglam/imagepicker/adapter/ImagePickerAdapter.java
package com.nguyenhoanglam.imagepicker.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.recyclerview.widget.RecyclerView;

import com.nguyenhoanglam.imagepicker.R;
import com.nguyenhoanglam.imagepicker.helper.ImageHelper;
import com.nguyenhoanglam.imagepicker.listener.OnImageClickListener;
import com.nguyenhoanglam.imagepicker.listener.OnImageSelectionListener;
import com.nguyenhoanglam.imagepicker.model.Config;
import com.nguyenhoanglam.imagepicker.model.Image;
import com.nguyenhoanglam.imagepicker.ui.common.BaseRecyclerViewAdapter;
import com.nguyenhoanglam.imagepicker.ui.imagepicker.ImageLoader;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Created by hoanglam on 7/31/16.
 */
public class ImagePickerAdapter extends BaseRecyclerViewAdapter<ImagePickerAdapter.ImageViewHolder> {

    private Config config;
    private List<Image> images = new ArrayList<>();
    private List<Image> selectedImages = new ArrayList<>();
    private OnImageClickListener itemClickListener;
    private OnImageSelectionListener imageSelectionListener;

    public ImagePickerAdapter(Context context, Config config, ImageLoader imageLoader, List<Image> selectedImages, OnImageClickListener itemClickListener) {
        super(context, imageLoader);
        this.config = config;
        this.itemClickListener = itemClickListener;

        if (selectedImages != null && !selectedImages.isEmpty()) {
            this.selectedImages.addAll(selectedImages);
        }
    }

    @Override
    public ImageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = getInflater().inflate(R.layout.imagepicker_item_image, parent, false);
        return new ImageViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final ImageViewHolder viewHolder, final int position) {

        final Image image = images.get(position);
        final int selectedPosition = getSelectedPosition(image);
        final boolean isSelected = selectedPosition != -1;

        getImageLoader().loadImage(image.getPath(), viewHolder.image);

        viewHolder.gifIndicator.setVisibility(ImageHelper.isGifFormat(image) ? View.VISIBLE : View.GONE);

        viewHolder.selectedIcon.setVisibility(isSelected && !config.isShowSelectedAsNumber() ? View.VISIBLE : View.GONE);
        viewHolder.selectedNumber.setVisibility(isSelected && config.isShowSelectedAsNumber() ? View.VISIBLE : View.GONE);
        if (viewHolder.selectedNumber.getVisibility() == View.VISIBLE) {
            viewHolder.selectedNumber.setText((selectedPosition + 1) + "");
        }

        viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean shouldSelect = itemClickListener.onImageClick(view, viewHolder.getAdapterPosition(), !isSelected);
                if (isSelected) {
                    removeSelected(image, position);
                } else if (shouldSelect) {
                    addSelected(image, position);
                } else {
                    String message = String.format(config.getLimitMessage(), config.getMaxSize());
                    Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private int getSelectedPosition(Image image) {
        for (int i = 0; i < selectedImages.size(); i++) {
            Image selectedImage = selectedImages.get(i);
            if (selectedImage.getPath().equals(image.getPath())) {
                return i;
            }
        }
        return -1;
    }

    public void setOnImageSelectionListener(OnImageSelectionListener imageSelectedListener) {
        this.imageSelectionListener = imageSelectedListener;
    }

    @Override
    public int getItemCount() {
        return images.size();
    }


    public void setData(List<Image> images) {
        if (images != null) {
            this.images.clear();
            this.images.addAll(images);
        }
        notifyDataSetChanged();
    }

    public void addSelected(List<Image> images) {
        selectedImages.addAll(images);
        notifySelectionChanged();
    }

    public void addSelected(Image image, int position) {
        selectedImages.add(image);
        notifyItemChanged(position);
        notifySelectionChanged();
    }

    public void removeSelected(Image image, int position) {
        Iterator<Image> itr = selectedImages.iterator();
        while (itr.hasNext()) {
            Image itrImage = itr.next();
            if (itrImage.getId() == image.getId()) {
                itr.remove();
                break;
            }
        }
        notifyDataSetChanged();
        notifySelectionChanged();
    }

    public void removeAllSelected() {
        selectedImages.clear();
        notifyDataSetChanged();
        notifySelectionChanged();
    }

    private void notifySelectionChanged() {
        if (imageSelectionListener != null) {
            imageSelectionListener.onSelectionUpdate(selectedImages);
        }
    }

    public List<Image> getSelectedImages() {
        return selectedImages;
    }

    static class ImageViewHolder extends RecyclerView.ViewHolder {

        private ImageView image;
        private ImageView selectedIcon;
        private TextView selectedNumber;
        private View gifIndicator;

        public ImageViewHolder(View itemView) {
            super(itemView);
            image = itemView.findViewById(R.id.image_thumbnail);
            selectedIcon = itemView.findViewById(R.id.image_selected_icon);
            selectedNumber = itemView.findViewById(R.id.text_selected_number);
            gifIndicator = itemView.findViewById(R.id.gif_indicator);

        }

    }

}

File: imagepicker/src/main/java/com/nguyenhoanglam/imagepicker/adapter/ImagePickerAdapter.kt
package com.nguyenhoanglam.imagepicker.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.nguyenhoanglam.imagepicker.R
import com.nguyenhoanglam.imagepicker.helper.ImageHelper
import com.nguyenhoanglam.imagepicker.listener.OnImageClickListener
import com.nguyenhoanglam.imagepicker.listener.OnImageSelectionListener
import com.nguyenhoanglam.imagepicker.model.Config
import com.nguyenhoanglam.imagepicker.model.Image
import com.nguyenhoanglam.imagepicker.ui.common.BaseRecyclerViewAdapter
import com.nguyenhoanglam.imagepicker.ui.imagepicker.ImageLoader
import java.util.*

/**
 * Created by hoanglam on 7/31/16.
 */
class ImagePickerAdapter(context: Context, private val config: Config, imageLoader: ImageLoader, selectedImages: List<Image>, private val itemClickListener: OnImageClickListener) : BaseRecyclerViewAdapter<ImagePickerAdapter.ImageViewHolder?>(context, imageLoader) {

    private val images: MutableList<Image> = ArrayList()
    private val selectedImages: MutableList<Image> = ArrayList()
    private var imageSelectionListener: OnImageSelectionListener? = null

    init {
        if (selectedImages != null && selectedImages.isNotEmpty()) {
            this.selectedImages.addAll(selectedImages)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ImageViewHolder {
        val itemView = inflater.inflate(R.layout.imagepicker_item_image, parent, false)
        return ImageViewHolder(itemView)
    }

    override fun onBindViewHolder(viewHolder: ImageViewHolder, position: Int) {
        val image = images[position]
        val selectedPosition = getSelectedPosition(image)
        val isSelected = selectedPosition != -1

        imageLoader.loadImage(image.path, viewHolder.image)
        viewHolder.gifIndicator.visibility = if (ImageHelper.isGifFormat(image)) View.VISIBLE else View.GONE
        viewHolder.selectedIcon.visibility = if (isSelected && !config.isShowSelectedAsNumber) View.VISIBLE else View.GONE
        viewHolder.selectedNumber.visibility = if (isSelected && config.isShowSelectedAsNumber) View.VISIBLE else View.GONE
        if (viewHolder.selectedNumber.visibility == View.VISIBLE) {
            viewHolder.selectedNumber.text = (selectedPosition + 1).toString()
        }
        viewHolder.itemView.setOnClickListener { view ->
            val shouldSelect = itemClickListener.onImageClick(view, viewHolder.adapterPosition, !isSelected)
            when {
                isSelected -> {
                    removeSelected(image, position)
                }
                shouldSelect -> {
                    addSelected(image, position)
                }
                else -> {
                    val message = String.format(config.limitMessage, config.maxSize)
                    Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun getSelectedPosition(image: Image): Int {
        for (i in selectedImages.indices) {
            val selectedImage = selectedImages[i]
            if (selectedImage.path == image.path) {
                return i
            }
        }
        return -1
    }

    fun setOnImageSelectionListener(imageSelectedListener: OnImageSelectionListener?) {
        imageSelectionListener = imageSelectedListener
    }

    override fun getItemCount(): Int {
        return images.size
    }

    fun setData(images: List<Image>?) {
        if (images != null) {
            this.images.clear()
            this.images.addAll(images)
        }
        notifyDataSetChanged()
    }

    fun addSelected(images: List<Image>?) {
        selectedImages.addAll(images!!)
        notifySelectionChanged()
    }

    private fun addSelected(image: Image, position: Int) {
        selectedImages.add(image)
        notifyItemChanged(position)
        notifySelectionChanged()
    }

    private fun removeSelected(image: Image, position: Int) {
        val itr = selectedImages.iterator()
        while (itr.hasNext()) {
            val itrImage = itr.next()
            if (itrImage.id == image.id) {
                itr.remove()
                break
            }
        }
        notifyDataSetChanged()
        notifySelectionChanged()
    }

    fun removeAllSelected() {
        selectedImages.clear()
        notifyDataSetChanged()
        notifySelectionChanged()
    }

    private fun notifySelectionChanged() {
        if (imageSelectionListener != null) {
            imageSelectionListener!!.onSelectionUpdate(selectedImages)
        }
    }

    fun getSelectedImages(): List<Image> {
        return selectedImages
    }

    class ImageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val image: ImageView = itemView.findViewById(R.id.image_thumbnail)
        val selectedIcon: ImageView = itemView.findViewById(R.id.image_selected_icon)
        val selectedNumber: TextView = itemView.findViewById(R.id.text_selected_number)
        val gifIndicator: View = itemView.findViewById(R.id.gif_indicator)

    }
}